### ISSUE TITLE
Training gets stuck at loss.backward() using TPU

### ISSUE THREAD

**Issue by evelynkyl** [huggingface/accelerate#283]
## ❓ Questions and Help

Hi, I am trying to fine-tune a pre-trained language model (~400M) using ParlAI and XLA via huggingface's accelerate. But the training gets stuck at `loss.backward()` step. One thing about the ParlAI is that they seperate the data loading and main model training into different scripts, therefore I have to use `self.model, self.optimizer, batch = accelerator.prepare(self.model, self.optimizer, batch)` multiple times otherwise the system screams that the tensors are not XLA tensors. I would really appreciate it If anyone can take a look and advise on how to fix this issue.

Below is the output of the metric report along with other print statements I used for debugging.
```
11:13:27 INFO | training...
the computed loss is tensor(1.9162, device='xla:1', grad_fn=<DivBackward0>)
performing accelerator.backward(loss):  None
this is torch_xla._XLAC._get_xla_tensors_text([loss] IR {
  %0 = s64[16]{0} xla::device_data(), device=TPU:0
  %1 = s64[] aten::sum(%0), dimensions=(0), keep_reduced_dimensions=0, dtype=4
  %2 = f32[] xla::cast(%1), type=f32, dtype=Float
  %3 = f32[16]{0} xla::device_data(), device=TPU:0
  %4 = f32[] aten::sum(%3), dimensions=(0), keep_reduced_dimensions=0, dtype=6
  %5 = f32[] aten::div(%4, %2), ROOT=0
}

this is a metric report Metric: CompileTime
  TotalSamples: 15
  Accumulator: 19s334ms940.755us
  ValueRate: 01s021ms185.841us / second
  Rate: 0.792274 / second
  Percentiles: 1%=007ms553.223us; 5%=007ms553.223us; 10%=007ms425.795us; 20%=009ms072.881us; 50%=021ms906.075us; 80%=998ms656.162us; 90%=09s994ms042.677us; 95%=09s047ms368.065us; 99%=09s047ms368.065us
Metric: DeviceLockWait
  TotalSamples: 15
  Accumulator: 053.267us
  ValueRate: 002.686us / second
  Rate: 0.756331 / second
  Percentiles: 1%=001.586us; 5%=001.586us; 10%=001.905us; 20%=002.947us; 50%=003.184us; 80%=005.166us; 90%=005.648us; 95%=005.716us; 99%=005.716us
Metric: ExecuteTime
  TotalSamples: 15
  Accumulator: 211ms826.316us
  ValueRate: 011ms134.990us / second
  Rate: 0.792239 / second
  Percentiles: 1%=002ms386.367us; 5%=002ms386.367us; 10%=002ms455.348us; 20%=002ms478.949us; 50%=003ms287.108us; 80%=014ms827.398us; 90%=071ms815.250us; 95%=073ms143.528us; 99%=073ms143.528us
Metric: InboundData
  TotalSamples: 18
  Accumulator: 2.41KB
  ValueRate: 130.39B / second
  Rate: 0.951754 / second
  Percentiles: 1%=1.00B; 5%=1.00B; 10%=1.00B; 20%=1.00B; 50%=8.00B; 80%=128.00B; 90%=128.00B; 95%=1.73KB; 99%=1.73KB
Metric: InputOutputAliasCount
  TotalSamples: 11
  Accumulator: 12.00
  ValueRate: 0.61 / second
  Rate: 0.55474 / second
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=1.00; 80%=1.00; 90%=1.00; 95%=2.00; 99%=2.00
Metric: IrValueTensorToXlaData
  TotalSamples: 357
  Accumulator: 15s110ms619.120us
  ValueRate: 424ms335.772us / second
  Rate: 10.0259 / second
  Percentiles: 1%=001ms069.136us; 5%=001ms279.980us; 10%=001ms413.787us; 20%=002ms551.990us; 50%=002ms841.347us; 80%=066ms132.470us; 90%=067ms457.525us; 95%=261ms053.985us; 99%=270ms630.855us
Metric: OutboundData
  TotalSamples: 368
  Accumulator: 1.36GB
  ValueRate: 38.75MB / second
  Rate: 10.2479 / second
  Percentiles: 1%=4.00B; 5%=5.00KB; 10%=5.00KB; 20%=5.00KB; 50%=5.00KB; 80%=6.25MB; 90%=6.25MB; 95%=25.00MB; 99%=25.00MB
Metric: ReleaseDataHandlesTime
  TotalSamples: 14
  Accumulator: 153ms995.227us
  ValueRate: 008ms089.200us / second
  Rate: 0.740211 / second
  Percentiles: 1%=738.550us; 5%=738.550us; 10%=962.790us; 20%=001ms031.819us; 50%=002ms699.409us; 80%=016ms770.251us; 90%=019ms555.057us; 95%=080ms454.166us; 99%=080ms454.166us
Metric: TensorsGraphSize
  TotalSamples: 15
  Accumulator: 13737.00
  ValueRate: 725.57 / second
  Rate: 0.792275 / second
  Percentiles: 1%=3.00; 5%=3.00; 10%=3.00; 20%=5.00; 50%=7.00; 80%=15.00; 90%=6823.00; 95%=6823.00; 99%=6823.00
Metric: TransferFromServerTime
  TotalSamples: 18
  Accumulator: 066ms155.154us
  ValueRate: 003ms497.967us / second
  Rate: 0.951754 / second
  Percentiles: 1%=001ms306.763us; 5%=001ms306.763us; 10%=001ms398.792us; 20%=002ms544.387us; 50%=002ms869.355us; 80%=008ms518.333us; 90%=012ms099.418us; 95%=013ms433.408us; 99%=013ms433.408us
Metric: TransferToServerTime
  TotalSamples: 368
  Accumulator: 15s131ms507.769us
  ValueRate: 421ms440.948us / second
  Rate: 10.2502 / second
  Percentiles: 1%=001ms063.388us; 5%=001ms270.664us; 10%=001ms390.961us; 20%=002ms546.362us; 50%=002ms829.677us; 80%=066ms102.141us; 90%=067ms432.965us; 95%=261ms927.295us; 99%=270ms621.466us
Metric: TransferToServerTransformTime
  TotalSamples: 368
  Accumulator: 389ms949.399us
  ValueRate: 011ms831.283us / second
  Rate: 10.2479 / second
  Percentiles: 1%=043.033us; 5%=050.050us; 10%=059.885us; 20%=087.111us; 50%=130.146us; 80%=002ms578.564us; 90%=002ms702.544us; 95%=006ms921.781us; 99%=008ms652.382us
Counter: CreateCompileHandles
  Value: 15
Counter: CreateDataHandles
  Value: 384
Counter: CreateXlaTensor
  Value: 5308
Counter: DestroyDataHandles
  Value: 15
Counter: DestroyXlaTensor
  Value: 4603
Counter: DeviceDataCacheMiss
  Value: 10
Counter: ReleaseDataHandles
  Value: 15
Counter: UncachedCompile
  Value: 15
Counter: XRTAllocateFromTensor_Empty
  Value: 13
Counter: XrtCompile_Empty
  Value: 384
Counter: XrtExecuteChained_Empty
  Value: 384
Counter: XrtExecute_Empty
  Value: 384
Counter: XrtMemoryInfo_Empty
  Value: 384
Counter: XrtRead_Empty
  Value: 384
Counter: XrtReleaseAllocationHandle_Empty
  Value: 384
Counter: XrtReleaseCompileHandle_Empty
  Value: 384
Counter: XrtSessionCount
  Value: 4
Counter: XrtSubTuple_Empty
  Value: 384
Counter: aten::_local_scalar_dense
  Value: 9
Counter: aten::cumsum
  Value: 1
Counter: aten::masked_select
  Value: 1
Counter: xla::_copy_from
  Value: 422
Counter: xla::_log_softmax
  Value: 1
Counter: xla::_log_softmax_backward_data
  Value: 1
Counter: xla::_softmax
  Value: 26
Counter: xla::_softmax_backward_data
  Value: 26
Counter: xla::_unsafe_view
  Value: 133
Counter: xla::abs
  Value: 2
Counter: xla::add
  Value: 135
Counter: xla::add_
  Value: 132
Counter: xla::addcmul
  Value: 42
Counter: xla::any
  Value: 1
Counter: xla::arange_out
  Value: 1
Counter: xla::as_strided
  Value: 382
Counter: xla::bernoulli_
  Value: 42
Counter: xla::bitwise_and_out
  Value: 1
Counter: xla::bmm
  Value: 156
Counter: xla::cat
  Value: 1
Counter: xla::ceil
  Value: 1
Counter: xla::clamp_
  Value: 1
Counter: xla::clone
  Value: 53
Counter: xla::cumsum
  Value: 1
Counter: xla::div
  Value: 28
Counter: xla::div_
  Value: 69
Counter: xla::embedding
  Value: 4
Counter: xla::embedding_dense_backward
  Value: 2
Counter: xla::empty
  Value: 446
Counter: xla::empty_strided
  Value: 382
Counter: xla::eq
  Value: 30
Counter: xla::expand
  Value: 44
Counter: xla::fill_
  Value: 14
Counter: xla::gelu
  Value: 14
Counter: xla::gelu_backward
  Value: 14
Counter: xla::gt
  Value: 2
Counter: xla::index_select
  Value: 4
Counter: xla::lt
  Value: 1
Counter: xla::masked_fill_
  Value: 52
Counter: xla::masked_select
  Value: 1
Counter: xla::max
  Value: 4
Counter: xla::min
  Value: 1
Counter: xla::mm
  Value: 399
Counter: xla::mul
  Value: 262
Counter: xla::mul_
  Value: 4
Counter: xla::native_batch_norm
  Value: 42
Counter: xla::native_batch_norm_backward
  Value: 42
Counter: xla::ne
  Value: 5
Counter: xla::nll_loss_backward
  Value: 1
Counter: xla::nll_loss_forward
  Value: 1
Counter: xla::repeat
  Value: 26
Counter: xla::select
  Value: 4
Counter: xla::slice
  Value: 8
Counter: xla::sub
  Value: 1
Counter: xla::sum
  Value: 221
Counter: xla::t
  Value: 532
Counter: xla::transpose
  Value: 416
Counter: xla::tril
  Value: 12
Counter: xla::unbind
  Value: 2
Counter: xla::unsqueeze
  Value: 17
Counter: xla::view
  Value: 1639
Counter: xla::zero_
  Value: 1
Metric: XrtAllocateFromTensor
  TotalSamples: 368
  Accumulator: 674ms548.126us
  Mean: 002ms830.294us
  StdDev: 003ms035.214us
  Rate: 10.2503 / second
  Percentiles: 25%=429.054us; 50%=554.600us; 80%=002ms440.790us; 90%=003ms026.023us; 95%=008ms286.670us; 99%=019ms699.977us
Metric: XrtCompile
  TotalSamples: 15
  Accumulator: 19s091ms002.111us
  Mean: 01s273ms733.474us
  StdDev: 03s034ms709.782us
  Rate: 0.792277 / second
  Percentiles: 25%=007ms006.559us; 50%=019ms110.326us; 80%=932ms812.838us; 90%=09s961ms819.611us; 95%=09s010ms994.333us; 99%=09s010ms994.333us
Metric: XrtExecute
  TotalSamples: 15
  Accumulator: 160ms538.076us
  Mean: 011ms635.872us
  StdDev: 023ms381.113us
  Rate: 0.792158 / second
  Percentiles: 25%=001ms106.660us; 50%=002ms532.117us; 80%=002ms086.252us; 90%=069ms194.319us; 95%=071ms271.787us; 99%=071ms271.787us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReadLiteral
  TotalSamples: 19
  Accumulator: 010ms266.051us
  Mean: 540.318us
  StdDev: 147.028us
  Rate: 1.00464 / second
  Percentiles: 25%=375.627us; 50%=565.235us; 80%=679.678us; 90%=714.152us; 95%=830.439us; 99%=830.439us
Metric: XrtReleaseAllocation
  TotalSamples: 14
  Accumulator: 283.906us
  Mean: 020.279us
  StdDev: 005.056us
  Rate: 0.740218 / second
  Percentiles: 25%=018.544us; 50%=021.219us; 80%=023.325us; 90%=024.486us; 95%=032.160us; 99%=032.160us
```

## To Reproduce
The main training scripts are as below.
[scripts.torch_generator_agent.txt](https://github.com/pytorch/xla/files/8252994/scripts.torch_generator_agent.txt)
[train_model.txt](https://github.com/pytorch/xla/files/8252992/train_model.txt)
[torch_agent.txt](https://github.com/pytorch/xla/files/8253003/torch_agent.txt)

Environment
----
reproducible on XLA backend [CPU/TPU]: tested on a colab TPU
torch_xla version: cloud-tpu-client==0.10 and torch_xla-1.9-cp37-cp37m-linux_x86_64
torch==1.9.0+cu111
I used pip install accelerate

Thank you so much in advance!

**Comment by sgugger** [comment]
Please use the [forums](https://discuss.huggingface.co/) to debug your code, as we keep the issues for bugs and feature requests only :-)

Not passing the whole dataloader will result in each process seeing the same data, which completely kills the purpose of distributed training. So you should find a way to import the dataloader from your other script.

**Comment by evelynkyl** [comment]
Cool, thanks for letting for me know! And sorry about posting it here. I will try to see implement dataloader if possible.
